brainglobe_atlasapi/atlas_generation/atlas_scripts/allen_bbp_ccfv3a_entire_mouse_brain.py

```diff
@@ -0,0 +1,267 @@
+
+
+def flatten_structure_tree(node, structure_id_path=None):
+    """Recursively flatten the tree structure."""
+    if structure_id_path is None:
+        structure_id_path = []
+
+    # Create current path
+    current_path = structure_id_path + [node["id"]]
+
+    # Create entry for current node
+    entry = {
+        "id": node["id"],
+        "name": node["name"],
+        "acronym": node["acronym"],
+        "structure_id_path": current_path,
+        "rgb_triplet": hex_to_rgb(node["color_hex_triplet"]),
+    }
+
+    # Start with current node's data
+    entries = [entry]
+
+    # Recursively process children
+    if "children" in node:
+        for child in node["children"]:
+            entries.extend(flatten_structure_tree(child, current_path))
+
+    return entries
+
+
+def retrieve_structure_information(download_path):
+    """
+    Return a pandas DataFrame with flattened atlas structure information.
+    """
+    with open(download_path / "hierarchy_bbp_atlas_pipeline.json") as f:
+        metadata_json = json.load(f)
+
+    # Get the root of the hierarchy
+    root = metadata_json["msg"][0]
+
+    # Flatten the tree structure
+    flattened_structures = flatten_structure_tree(root)
+
+    return flattened_structures
+
+
+def retrieve_or_construct_meshes(download_path, annotated_volume, structures):
+    """
+    This function should return a dictionary of ids and corresponding paths to
+    mesh files. Some atlases are packaged with mesh files, in these cases we
+    should use these files. Then this function should download those meshes.
+    In other cases we need to construct the meshes ourselves. For this we have
+    helper functions to achieve this.
+    """
+    meshes_dict = construct_meshes_from_annotation(
+        download_path, annotated_volume, structures, METADATA.root_id
+    )
+    return meshes_dict
+
+
+### If the code above this line has been filled correctly, nothing needs to be
+### edited below (unless variables need to be passed between the functions).
+if __name__ == "__main__":
+    bg_root_dir = Path.home() / "brainglobe_workingdir" / METADATA.name
+    bg_root_dir.mkdir(exist_ok=True)
+    download_resources(
+        download_dir_path=bg_root_dir,
+        atlas_name=METADATA.name,
+        atlas_file_url=METADATA.atlas_link,
+    )
+    for resolution in [25, 10]:
+        reference_volume, annotated_volume = retrieve_reference_and_annotation(
+            bg_root_dir, resolution=resolution
+        )
+        additional_references = retrieve_additional_references(
+            bg_root_dir, resolution=resolution
+        )
+
+        hemispheres_stack = retrieve_hemisphere_map()
+        if resolution == 25:
+            structures = retrieve_structure_information(bg_root_dir)
+            meshes_dict = retrieve_or_construct_meshes(
+                bg_root_dir, annotated_volume, structures
+            )
+        reference_volume = (reference_volume * 65535).astype(np.uint16)
```

Comment:
presumably you're sure the reference volume pixels are all between 0.0 and 1.0 here?

Reply:
Yeah :)